arboretum/kubernetes/fetchers/fetch_cluster_resource.py

```diff
@@ -0,0 +1,159 @@
+        :param str verify: Path for CA certificate file, or None if skip TLS
+                           server certificate verification (same as "curl -k")
+        :param list(str) resource_types: list of resource type in plural form
+                                         (e.g., "pods") or custom resource API
+                                         name (e.g.,
+                                         "apigroup.example.com/v1/mycustom")
+        """
+        s = self.session(server_url)
+        s.headers.update({'Authorization': f'Bearer {token}'})
+        resources = {}
+        for resource_type in resource_types:
+            if '/' in resource_type:
+                url = f'{server_url}/apis/{resource_type}'
+            else:
+                url = f'{server_url}/api/v1/{resource_type}'
+            resp = s.get(url, verify=verify)
+            if resp.status_code == 404:
+                self.logger.warning(
+                    'Cluster "%s" does not provide resource type "%s"',
+                    server_url,
+                    resource_type
+                )
+                continue
+            resp.raise_for_status()
+            resources[resource_type] = resp.json()['items']
+        return resources
+
+    @store_raw_evidence('kubernetes/cluster_resource.json')
+    def fetch_cluster_resource(self):
+        """Fetch cluster resources of listed clusters."""
+        ev = get_evidence_dependency(
+            'raw/auditree/compliance_config.json', self.locker
+        )
+        config = json.loads(ev.content)['org']['kubernetes']
+
+        cluster_list = config['cluster_list']
+        self.logger.debug('number of clusters: %d', len(cluster_list))
+
+        resource_types = self.RESOURCE_TYPE_DEFAULT
+        if ('cluster_resource' in config
+                and 'target_resource_types' in config['cluster_resource']):
+            resource_types = (
+                config['cluster_resource']['target_resource_types']
+            )
+        resources = {}
+
+        for c in cluster_list:
+            self.logger.info('Getting resource from cluster: %s', c['name'])
+            try:
+                with open(c['kubeconfig']) as f:
+                    kubeconfig = yaml.safe_load(f)
+            except OSError:
+                self.logger.error(
+                    'Cannot access cluster "%s" because '
+                    'kubeconfig file is not found: %s',
+                    c['serverURL'],
+                    c['kubeconfig']
+                )
+                raise
+            for cluster in kubeconfig['clusters']:
+                ca_filepath = cluster['cluster']['certificate-authority']
+                server_url = cluster['cluster']['server']
+                cluster_name = cluster['name']
+                if cluster_name != c['name']:
+                    self.logger.info(
+                        'Skip non-target cluster: %s', cluster_name
+                    )
+                    continue
```

Comment:
Is the implication here that there will be multiple clusters in `kubeconfig['clusters']` where the name matches `c['name']`?

Assuming for the moment that the answer to my question is yes, then...

If you're looking to skip clusters it would be better to perform the skip logic upfront before you start performing operations.

```suggestion
            kc_clusters = kubeconfig['clusters']
            for cluster in filter(lambda kc: kc == c['name'], kc_clusters):
                ca_filepath = clstr['cluster']['certificate-authority']
                server_url = clstr['cluster']['server']
```

I also don't see a reason not to just continue to reference the cluster name as `cluster['name']` or even `c['name']`.  It doesn't really buy you anything to assign it to the `cluster_name` variable.

Reply:
I experimented the behaviour of `kubectl config` and the result shows that the answer to your question is "no".
I think your suggestion also works even if the answer is "no"; do you have much better solution for this?

The following code can remove `for`, but for me your suggestion is simpler because yours does not need the zero element check.

```python
target_clusters = [kc for kc in kc_clusters if kc['name'] == c['name']]
if len(target_clusters) == 0:
    continue
ca_filepath = target_clusters[0]['cluster']['certificate-authority']
...
```

Reply:
changed to this.

```suggestion
            kc_clusters = kubeconfig['clusters']
            for cluster in filter(lambda kc: kc['name'] == c['name'],
                                  kc_clusters):
                ca_filepath = cluster['cluster']['certificate-authority']
                server_url = cluster['cluster']['server']

```